File: tlg/src/main/java/ch/bailu/tlg/PlatformContext.java
package ch.bailu.tlg;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;


public abstract class PlatformContext
{
    public void setDirtyRect(TlgRectangle rect) {
        /*TetrisRectangle r = new TetrisRectangle(rect);

        r.grow(1);
        drawRectangle(colorGrid(), r);*/
    }

    public void drawRectangle(int c, TlgRectangle r) {
        drawLine(c, r.getTL(), r.getTR());
        drawLine(c, r.getTR(), r.getBR());
        drawLine(c, r.getBR(), r.getBL());
        drawLine(c, r.getBL(), r.getTL());
    }

    public abstract void drawLine(int color, TlgPoint p1, TlgPoint p2);
    public abstract void drawFilledRectangle(int color, TlgRectangle rect);
    public abstract void drawText(int color, TlgRectangle rect, String text);

    // TODO: move color stuff out of context
    public abstract int colorBackground();
    public abstract int colorDark();
    public abstract int colorHighlight();
    public abstract int colorGrayed();
    public abstract int colorFrame();
    public abstract int colorGrid();

    public abstract int countOfColor();
    public abstract int getColor(int i);
    //////

    public File getConfigDirectory() {
        return new File(System.getProperty("user.home"), ".config/tlg");
    }
    public abstract void onNewHighscore();

    public BufferedInputStream  getInputStream(String fileName) throws FileNotFoundException {
        File file = new File(getConfigDirectory(), fileName);
        return new BufferedInputStream( new FileInputStream(file));
    }

    public BufferedOutputStream getOutputStream(String fileName) throws FileNotFoundException {
        File directory = getConfigDirectory();
        directory.mkdirs();
        File file = new File(directory, fileName);
        return new BufferedOutputStream( new FileOutputStream(file));
    }
}

File: tlg/src/main/kotlin/context/PlatformContext.kt
package context

import geometry.TlgPoint
import geometry.TlgRectangle
import java.io.*

abstract class PlatformContext {
    open fun setDirtyRect(rect: TlgRectangle) {
        /*TetrisRectangle r = new TetrisRectangle(rect);

        r.grow(1);
        drawRectangle(colorGrid(), r);*/
    }

    fun drawRectangle(c: Int, r: TlgRectangle) {
        drawLine(c, r.tL, r.tR)
        drawLine(c, r.tR, r.bR)
        drawLine(c, r.bR, r.bL)
        drawLine(c, r.bL, r.tL)
    }

    abstract fun drawLine(color: Int, p1: TlgPoint, p2: TlgPoint)
    abstract fun drawFilledRectangle(color: Int, rect: TlgRectangle)
    abstract fun drawText(color: Int, rect: TlgRectangle, text: String)

    // TODO: move color stuff out of context
    abstract fun colorBackground(): Int
    abstract fun colorDark(): Int
    abstract fun colorHighlight(): Int
    abstract fun colorGrayed(): Int
    abstract fun colorFrame(): Int
    abstract fun colorGrid(): Int
    abstract fun countOfColor(): Int
    abstract fun getColor(i: Int): Int

    //////
    open val configDirectory: File
        get() = File(System.getProperty("user.home"), ".config/tlg")

    abstract fun onNewHighscore()
    @Throws(FileNotFoundException::class)
    fun getInputStream(fileName: String): BufferedInputStream {
        val file = File(configDirectory, fileName)
        return BufferedInputStream(FileInputStream(file))
    }

    @Throws(FileNotFoundException::class)
    fun getOutputStream(fileName: String): BufferedOutputStream {
        val directory = configDirectory
        directory.mkdirs()
        val file = File(directory, fileName)
        return BufferedOutputStream(FileOutputStream(file))
    }
}
